File: lib/util/wifi_utils.dart
import 'dart:math';

int calculateSignalLevel(int rssi) {
  const int MIN_RSSI = -100;
  const int MAX_RSSI = -55;
  int numLevels=5;
  if (rssi <= MIN_RSSI) {
    return 0;
  } else if (rssi >= MAX_RSSI) {
    return numLevels - 1;
  } else {
    return ((rssi - MIN_RSSI) * (numLevels - 1) / (MAX_RSSI - MIN_RSSI)).floor();
  }
}

num calculateDistance(int frequency, int level) {
  const DISTANCE_MHZ_M = 27.55; // Adjust the constant value as needed.
  return pow(10.0, (DISTANCE_MHZ_M - 20 * log(frequency.toDouble()) / log(10) + level.abs()) / 20.0);
}

File: android/app/src/main/kotlin/com/androidsutra/wifi_signal_monitor/model/WiFiUtils.kt

package com.androidsutra.wifi_signal_monitor.model

import com.androidsutra.wifi_signal_monitor.utils.EMPTY
import java.net.InetAddress
import java.nio.ByteOrder
import kotlin.math.abs
import kotlin.math.log10
import kotlin.math.pow

private const val DISTANCE_MHZ_M = 27.55
private const val MIN_RSSI = -100
private const val MAX_RSSI = -55
private const val QUOTE = "\""

fun calculateDistance(frequency: Int, level: Int): Double =
    10.0.pow((DISTANCE_MHZ_M - 20 * log10(frequency.toDouble()) + abs(level)) / 20.0)

fun calculateSignalLevel(rssi: Int, numLevels: Int): Int = when {
    rssi <= MIN_RSSI -> 0
    rssi >= MAX_RSSI -> numLevels - 1
    else -> (rssi - MIN_RSSI) * (numLevels - 1) / (MAX_RSSI - MIN_RSSI)
}

fun convertSSID(ssid: String): String = ssid.removePrefix(QUOTE).removeSuffix(QUOTE)

fun convertIpV4Address(ipV4Address: Int): String {
    return try {
        val value: Long = when (ByteOrder.LITTLE_ENDIAN) {
            ByteOrder.nativeOrder() -> Integer.reverseBytes(ipV4Address).toLong()
            else -> ipV4Address.toLong()
        }
        return InetAddress.getByAddress(value.toBigInteger().toByteArray()).hostAddress ?: String.EMPTY
    } catch (e: Exception) {
        String.EMPTY
    }
}
